### PR TITLE
Add sequential symbolic LU factorization

### LINKED ISSUE #1163: Symbolic LU factorization
We only have a symbolic Cholesky factorization ATM, which runs hybrid between CPU and device. We also need an LU factorization, which would probably only run on the CPU for now, but should be parallelized later.

### PR DESCRIPTION
**Pull request by upsj**
This adds a sequential LU factorization happening in `core` to support non-symmetric matrices.

Closes #1163

### CODE REVIEW

**Review by fritzgoebel — APPROVED** [review]
LGTM!

File: core/test/utils/assertions.hpp [hunk 1]
```diff
@@ -202,6 +202,86 @@ void print_columns(Ostream& os, const Iterator& begin, const Iterator& end)
...
+                }
+            }
+        }
+        os << (row % 10) << '\n';
+    }
+    os << ' ';
+    for (int col = 0; col < first.size[1]; col++) {
+        os << (col % 10);
+    }
+    os << "\n| is first, - is second, + is both\n";
```
Comment (thoasm) [inline core/test/utils/assertions.hpp:245]: Maybe mention that space is neither?
```suggestion
    os << "\n'|' is first, '-' is second, '+' is both, ' ' is neither\n";
```

File: core/test/utils/assertions.hpp [hunk 2]
```diff
@@ -763,6 +825,13 @@ ::testing::AssertionResult matrices_equal_sparsity(
 
     first_data.ensure_row_major_order();
     second_data.ensure_row_major_order();
+    // make sure if we write data to disk, it only contains ones.
+    for (auto& entry : first_data.nonzeros) {
+        entry.value = gko::one<typename LinOp1::value_type>();
```
Comment (thoasm) [inline core/test/utils/assertions.hpp:830]:
```suggestion
        entry.value = one<typename LinOp1::value_type>();
```

File: core/test/utils/assertions.hpp [hunk 3]
```diff
@@ -763,6 +825,13 @@ ::testing::AssertionResult matrices_equal_sparsity(
...
+    }
+    for (auto& entry : second_data.nonzeros) {
+        entry.value = gko::one<typename LinOp2::value_type>();
```
Comment (thoasm) [inline core/test/utils/assertions.hpp:833]:
```suggestion
        entry.value = one<typename LinOp2::value_type>();
```

**Review by thoasm — APPROVED** [review]
LGTM!
I like the idea of `|` and `-` combined is `+` (for the sparsity pattern printing).

File: core/factorization/lu.cpp [hunk 4]
```diff
@@ -95,7 +95,7 @@ std::unique_ptr<LinOp> Lu<ValueType, IndexType>::generate_impl(
         if (parameters_.symmetric_sparsity) {
             factors = gko::factorization::symbolic_cholesky(mtx.get());
         } else {
-            GKO_NOT_SUPPORTED(mtx);
+            factors = gko::factorization::symbolic_lu(mtx.get());
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/factorization/lu.cpp:98]: I think it was necessary because `Lu` is in `gko::experimental::factorization`

**Comment by upsj** [comment]
rebase!

**Pull request merged**